### ISSUE TITLE
Does Jennifer strive for ActiveRecord method parity?

### ISSUE THREAD

**Issue by SebastianSzturo** [imdrasil/jennifer.cr#405]
In my project I am monkey-patching some native ActiveRecord methods like:

* `find_by`
* `find_by_and_initialize`
* `find_or_create_by`
* `Model.first` instead of `Model.all.first`

Is the long-term vision for Jennifer to be as close as possible to AR's API or do you want to be more selective? Just want to clarify before opening any PRs. 😄

**Comment by imdrasil** [comment]
well this is a bit complicated question. I'd like to accept a PR adding `#find_by`, `#find_or_initialize_by` and `#find_or_create_by`. Also it would be great to have `#first_or_create` and `#first_or_initialize` - but they may be a bit harder to implement. But for "Model.first instead of Model.all.first" I don't  have a clear vision for now.

The main philosophy behind forcing `.all` method when developer write a request is to separate query logic from business one that may leave in model. So all (well, almost all) query-related methods are available only from `Jennifer::QueryBuilder::ModelQuery` instance of which is instantiated by `Jennifer::Resource.all`. I need time to think about this and open for any thought.

To summarize it would be great to see PR for:
-  `Jennifer::Model::Base#find_or_initialize_by`
-  `Jennifer::Model::Base#find_or_create_by`
-  `Jennifer::Model::Base#find_or_create_by!`
- `Jennifer::QueryBuilder::IModelQuery#.find_by`
- `Jennifer::QueryBuilder::IModelQuery#.find_by!`

**Comment by imdrasil** [comment]
this is addressed [here](https://github.com/imdrasil/jennifer.cr/pull/420)
